sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -354,9 +355,14 @@ resource
     : identifier STRING
     ;
 
+insertStatement
+    : (ctes)? insertInto queryTerm queryOrganization                                       #singleInsertQuery
+    | (ctes)? fromClause multiInsertQueryBody+                                             #multiInsertQuery
```

Comment:
shall we change the `insertInto` and make the `insertInto` non-optional?

Reply:
@cloud-fan did you mean make insertInto non-optional in `multiInsertQueryBody` ? If so, i did try it and we need it to be optional to parse statement referenced in [test](https://github.com/apache/spark/blob/1b9fd67904671ea08526bfb7a97d694815d47665/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala#L133-L135)
Also , we we can have a mixture of insert and selects in the multiInsertQueryBody. Thats why we have it under a "multi insert" rule where insert is optional.

Reply:
why do we still need https://github.com/apache/spark/pull/24150/files#diff-8c1cb2af4aa1109e08481dae79052cc3R365 ?

Reply:
@cloud-fan If you don't mind, could you please paste the block or line here ? When i click on that link, it shows me pretty much all the changes so i am not sure which one i should focus on.

Reply:
@dilipbiswal . He means line 365. And, that's the same question with the below comment, https://github.com/apache/spark/pull/24150/files#r267283726.
```
| fromClause querySpecification queryOrganization                           #queryWithFrom
```

Line 365 is required for `FROM ... SELECT ...` statement.

Reply:
@dongjoon-hyun Thanks for helping out. When i clicked.. it took me to AstBuilder ... so i was confused.

@cloud-fan So we need the rule to parse statement i mentioned earlier.
```
from (from test select transform(....) using .. as ...) T select ...
```
It is parsed sort of recursively in the `queryNoWith` rule .. calling `queryNoWith` from `relationPrimary:aliasedQuery` rule.

Reply:
`fromClause multiInsertQueryBody+` -> `fromClause (insertInto? querySpecification queryOrganization)+` -> `fromClause (querySpecification queryOrganization)+`

and in `queryNoWith`, we have `fromClause querySpecification queryOrganization`. Seems we should change it to `fromClause (querySpecification queryOrganization)+`

Reply:
@cloud-fan Currently after this change, fromClause and multiple querySpecification queryOrganization will be handled by 2nd alternative of the insertStatement. I know that this may look confusing. In this pass, i wanted to handle single queries in one rule which is `query`. This would help us catch quite a few errors i.e when a user could plugin an insert in any part of the query. The rest is currently handled by `insertStatement` which includes multi insert, multi select and multi insert + select etc.  Please let me know what you think.

Reply:
@cloud-fan instead of name `insertStatement`, should we change to `insertOrMultiQueries` to clear the confusion ?